src/app/session/local-session.ts

```diff
@@ -0,0 +1,130 @@
+ * - Provide an interface to access the data
+ */
+
+import PouchDB from 'pouchdb-browser';
+
+import { Injectable } from '@angular/core';
+
+import { AppConfig } from '../app-config/app-config';
+import { User } from '../user/user';
+
+import { SyncState } from './sync-state.enum';
+import { LoginState } from './login-state.enum';
+import { StateHandler } from './util/state-handler';
+import { EntitySchemaService } from 'app/entity/schema/entity-schema.service';
+import { AlertService } from 'app/alerts/alert.service';
+
+@Injectable()
+export class LocalSession {
+  public database: any;
+  public liveSyncHandle: any;
+
+  public loginState: StateHandler<LoginState>; // logged in, logged out, login failed
+  public syncState: StateHandler<SyncState>; // started, completed, failed, unsynced
+  public currentUser: User;
+
+  constructor(private _alertService: AlertService, private _entitySchemaService: EntitySchemaService) {
+    this.database = new PouchDB(AppConfig.settings.database.name);
+
+    this.loginState = new StateHandler<LoginState>(LoginState.loggedOut);
+    this.syncState = new StateHandler<SyncState>(SyncState.unsynced);
+  }
+
+  /**
+   * Get a login at the local session by fetching the user from the local database and validating the password.
+   * Returns a Promise resolving with the loginState.
+   * Attention: This method waits for the first synchronisation of the database (or a fail of said initial sync).
+   * @param username Username
+   * @param password Password
+   */
+  public async login(username: string, password: string): Promise<LoginState> {
+    try {
+      await this.waitForFirstSync();
+      const userEntity = await this.loadUser(username);
+      if (userEntity.checkPassword(password)) {
+        this.loginState.setState(LoginState.loggedIn);
+        this.currentUser = userEntity;
+        return LoginState.loggedIn;
+      } else {
+        this.loginState.setState(LoginState.loginFailed);
+        return LoginState.loginFailed;
+      }
+    } catch (error) {
+      // possible error: initial sync failed
+      if (error && error.toState && error.toState === SyncState.failed) {
+        this._alertService.addDanger('The initial Sync of the Database failed, so you couldn\'t be logged in. Please try again later.');
+        this.loginState.setState(LoginState.loggedOut);
+        return LoginState.loggedOut;
+      }
+      // possible error: user object not found locally, which should return loginFailed.
+      if (error && error.status && error.status === 404) {
+        this.loginState.setState(LoginState.loginFailed);
+        return LoginState.loginFailed;
+      }
+      // all other cases must throw an error
+      throw error;
+    }
+  }
+
+  /**
+   * Wait for the first sync of the database, returns a Promise.
+   * Resolves directly, if the database is not initial, otherwise waits for the first change of the SyncState to completed (or failed)
+   */
+  public async waitForFirstSync() {

```

Comment:
Can we sketch an activity diagram to make the states entirely clear? Even a snap of a drawing would help a lot as a starting point.

Reply:
I added sequence diagrams and hope they help